### PR TITLE
Support Poetry version syntax and optional dependencies

### LINKED ISSUE #440: [BUG] Optional pyproject deps have optional added to the version
**Describe the bug**

Deps that are optional are written as _name versionoptional_

**Expected behavior**

The optional should not be added

**Outputs**
If applicable, add the output to help explain your problem.

Ref: https://github.com/conda-forge/aws-lambda-powertools-feedstock/pull/63

**Environment:**
 - Conda-forge update bot

**Additional context**

Change: https://github.com/conda-forge/aws-lambda-powertools-feedstock/pull/63/files

### LINKED ISSUE #441: it is not handling ~ in dependencies
Since the change you linked also contains this bug, I'll mention it explicitly: If dependencies in `pyproject.toml` use the version range shorthand (`~` or `^`), it does no conversion when populating the `meta.yaml` file. So, when presented with a dependency like so (using boto3 as an example):

```
boto3 = "~1.24"
```

it will produce
```
 - boto3 ~1.24
```

resulting in this error message
```
conda.exceptions.InvalidVersionSpec: Invalid version '~1.24': regex specs must start with '^' and end with '$'
```

---
It should instead produce the following
```
- boto3 >=1.24,<1.25
```

_Originally posted by @AlmightyYakob in https://github.com/conda-incubator/grayskull/issues/440#issuecomment-1412840767_

### PR DESCRIPTION
**Pull request by dlqqq**
@marcelotrevisani Thank you for all of your contributions! This is a critical project to the Conda Forge ecosystem, and I noticed that you have been maintaining this for 3+ years. I figured I should pay back your generosity with my own. I would appreciate your review on this PR. 🤗 ❤️ 

# Description

This PR implements support for the most common Poetry projects by correctly handling Poetry version syntax and optional dependencies.

- Closes #440
- Closes #441

What this PR implements:

- Correct encoding of Poetry dependency versions including support for the Poetry-exclusive `^` and `~` version operators
- Fixes bug where Poetry dependency specification fields like `optional` were being included in the Conda recipe version specifiers
- Support for Poetry optional dependencies declared via `optional = true`
- Implements additional unit and snapshot tests for Poetry projects

What this PR does NOT implement:

- Guaranteed correct handling of pre/post/dev release segments like `1.2.3a1`
- Support for Poetry Python restricted dependencies declared via `python = "..."`
- Support for all other Poetry dependency fields like: `git`, `path`, `url`, `source`, `markers`

# Strategy

## Version specifiers

Poetry version operators are mostly compatible with Conda recipe version operators, with the exception of `^` and `~` which need to have custom encoding logic. The expected behavior is roughly:

- `^M.m.p`⇒ `>=M.m.p,<(M+1).0.0`
- `~M.m.p`⇒ `>=M.m.p,<M.(m+1).0`

(format: Poetry version syntax ⇒ Conda recipe version syntax)

As you might expect, this logic is a bit trickier because Poetry also allows non-semver version specifiers like `~1` or `^0`, and has special edge cases for major version 0. Please see the implementation in `py_toml.py` for details.

## Optional dependencies

Optional dependencies need to be listed under `requirements.run_constrained` in the recipe metadata.

## Complete example

The Poetry dependency:

```
tensorflow-text = {version = "^2.11.0", optional = true, python = "^3.10, <3.12"}
```

Should be listed in the recipe metadata as:

```
requirements:
  run_constrained:
    tensorflow-text >=2.11.0,<3.0.0
```

We are implicitly ignoring all Poetry dependency specification fields aside from `version` and `optional` at the moment, as it’s not clear how to support them.

# Suggested next steps

- Review, iterate, and merge this PR.
- Cut a new 2.3.0 release.
- Investigate how to support [Python restricted dependencies](https://python-poetry.org/docs/dependency-specification#python-restricted-dependencies).
    - Initially I thought we could leverage [Conda recipe preprocessing selectors](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#preprocessing-selectors), but there is too much ambiguity on how to handle Python 3.10+, and it seems to be on a deprecation path. See upstream issue: [https://github.com/conda/conda-build/issues/4032](https://github.com/conda/conda-build/issues/4032)
- Perform a refactor of the code handling Conda recipe metadata synthesis. Let me explain my reasoning:
    - The source code is overly imperative, especially for the pyproject case. I had a lot of difficulty understanding the code because each metadata field was being determined manually through a long chain of dict lookups, dict merges, and conditional `if` statements. This style of code is brittle and prone to failure.
    - For example, the pyproject metadata gets merged manually with the setup.py/setup.cfg metadata, which then gets merged with PyPi metadata. This mysterious, Frankenstein "metadata" dictionary then finally gets transformed into Conda recipe metadata. This strategy will grow to be too difficult to maintain.
    - Eventually, I think the way forward is to have a `CondaMetadata` class that accepts a list of metadata sources and handles field resolution (i.e. “generate the list `requirements.run`, merging all dependencies together”) declaratively through a clean API.
    - @marcelotrevisani I can elaborate on this further in a future design document, if you are interested.

# References

- [`requirements.run_constrained`](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#run-constrained) in recipe metadata
- [Conda version syntax](https://conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#package-match-specifications)
- [Poetry version syntax](https://python-poetry.org/docs/dependency-specification)
- [Conda preprocessing selectors](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#preprocessing-selectors)

### CODE REVIEW

**Comment by marcelotrevisani** [comment]
Hi @dlqqq , thanks a lot for your contribution! I am happy that you are enjoying grayskull.

I have a few suggestions that I will add later today, I am at my day job, so I don't have much time to review right now, but I will do it later. Thank you!


🚀

**Comment by marcelotrevisani** [comment]
can you add ``semver`` to this file please?
https://github.com/conda/grayskull/blob/main/environment.yaml

**Comment by dlqqq** [comment]
@marcelotrevisani Done. 😎

File: tests/test_poetry.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    assert parse_version("1") == {"major": 1, "minor": None, "patch": None}
+    assert parse_version("1.2") == {"major": 1, "minor": 2, "patch": None}
+    assert parse_version("1.2.3") == {"major": 1, "minor": 2, "patch": 3}
+
+    with pytest.raises(InvalidVersion):
+        parse_version("asdf")
+    with pytest.raises(InvalidVersion):
+        parse_version("")
+    with pytest.raises(InvalidVersion):
+        parse_version(".")
```
Comment (marcelotrevisani) [inline tests/test_poetry.py:31]: can we divide this test in two parametrized tests please?
Something like
```suggestion
@pytest.mark.pararetrize(
    "version, major, minor, patch",
    [
        ("0", 0, None, None),
        ("1", 1, None, None),
        ("1.2", 1, 2, None),
        ("1.2.3", 1, 2, 3),
    ]
)
def test_parse_version_success(version, major, minor, patch):
    assert parse_version(version) == {"major": major, "minor": minor, "patch": patch}


@pytest.mark.parametrize("invalid_version", ["asdf", "", "."])
def test_parse_version_failure(invalid_version):
    with pytest.raises(InvalidVersion):
        parse_version(invalid_version)
```

File: tests/test_poetry.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+def test_get_caret_ceiling():
+    # examples from Poetry docs
+    assert get_caret_ceiling("0") == "1.0.0"
+    assert get_caret_ceiling("0.0") == "0.1.0"
+    assert get_caret_ceiling("0.0.3") == "0.0.4"
+    assert get_caret_ceiling("0.2.3") == "0.3.0"
+    assert get_caret_ceiling("1") == "2.0.0"
+    assert get_caret_ceiling("1.2") == "2.0.0"
+    assert get_caret_ceiling("1.2.3") == "2.0.0"
```
Comment (marcelotrevisani) [inline tests/test_poetry.py:42]:
```suggestion
@pytest.mark.parametrize(
    "version, ceiling_version",
    [
        ("0", "1.0.0"),
        ("0.0", "0.1.0"),
        ("0.0.3", "0.0.4"),
        ("0.2.3", "0.3.0"),
        ("1", "2.0.0"),
        ("1.2", "2.0.0"),
        ("1.2.3", "2.0.0"),
    ]
)
def test_get_caret_ceiling(version, ceiling_version):
    # examples from Poetry docs
    assert get_caret_ceiling(version) == ceiling_version
```

File: tests/test_poetry.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+
+def test_get_tilde_ceiling():
+    # examples from Poetry docs
+    assert get_tilde_ceiling("1") == "2.0.0"
+    assert get_tilde_ceiling("1.2") == "1.3.0"
+    assert get_tilde_ceiling("1.2.3") == "1.3.0"
```
Comment (marcelotrevisani) [inline tests/test_poetry.py:49]:
```suggestion
@pytest.mark.parametrize(
    "version, ceiling_version",
    [("1", "2.0.0"), ("1.2", "1.3.0"), ("1.2.3", "1.3.0")],
)
def test_get_tilde_ceiling(version, ceiling_version):
    # examples from Poetry docs
    assert get_tilde_ceiling(version) == ceiling_version
```

File: tests/test_poetry.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    assert encode_poetry_version("^0.2.3") == ">=0.2.3,<0.3.0"
+    assert encode_poetry_version("^1") == ">=1.0.0,<2.0.0"
+    assert encode_poetry_version("^1.2") == ">=1.2.0,<2.0.0"
+    assert encode_poetry_version("^1.2.3") == ">=1.2.3,<2.0.0"
+
+    # handle tilde operator correctly
+    # examples from Poetry docs
+    assert encode_poetry_version("~1") == ">=1.0.0,<2.0.0"
+    assert encode_poetry_version("~1.2") == ">=1.2.0,<1.3.0"
+    assert encode_poetry_version("~1.2.3") == ">=1.2.3,<1.3.0"
```
Comment (marcelotrevisani) [inline tests/test_poetry.py:80]:
```suggestion
@pytest.mark.parametrize(
    "version, encoded_version",
    [
        # should be unchanged
        ("1.*", "1.*"),
        (">=1,<2", ">=1,<2"),
        ("==1.2.3", "==1.2.3"),
        ("!=1.2.3", "!=1.2.3"),
        # strip spaces
        (">= 1, < 2", ">=1,<2"),
        # handle exact version specifiers correctly
        ("1.2.3", "1.2.3"),
        # handle caret operator correctly
        # examples from Poetry docs
        ("^0", ">=0.0.0,<1.0.0"),
        ("^0.0", ">=0.0.0,<0.1.0"),
        ("^0.0.3", ">=0.0.3,<0.1.0"),
        ("^0.2.3", ">=0.2.3,<0.3.0"),
        ("^1", ">=1.0.0,<2.0.0"),
        ("^1.2", ">=1.2.0,<2.0.0"),
        ("^1.2.3", ">=1.2.3,<2.0.0"),
        # handle tilde operator correctly
        # examples from Poetry docs
        ("~1", ">=1.0.0,<2.0.0"),
        ("~1.2", ">=1.2.0,<1.3.0"),
        ("~1.2.3", ">=1.2.3,<1.3.0"),
    ]
)
def test_encode_poetry_version(version, encoded_version):
    assert encode_poetry_version(version) == encoded_version
```

File: grayskull/strategy/pypi.py [hunk 5]
```diff
@@ -382,10 +383,14 @@ def get_metadata(recipe, config) -> dict:
     requirements_section = extract_requirements(metadata, config, recipe)
     optional_requirements = extract_optional_requirements(metadata, config)
     for key in requirements_section:
+        log.error(key)
```
Comment (marcelotrevisani) [inline grayskull/strategy/pypi.py:386]: Is it a left over?

File: grayskull/strategy/pypi.py [hunk 6]
```diff
@@ -571,6 +576,12 @@ def extract_requirements(metadata: dict, config, recipe) -> Dict[str, List[str]]
             "run": rm_duplicated_deps(sort_reqs(map(lambda x: x.lower(), run_req))),
         }
     )
+
+    if (
+        "requirements_run_constrained" in metadata
+        and metadata["requirements_run_constrained"]
+    ):
```
Comment (marcelotrevisani) [inline grayskull/strategy/pypi.py:583]:
```suggestion
    if metadata.get("requirements_run_constrained", None):
```

File: tests/test_poetry.py (hunk 1 above)
Comment (dlqqq) [reply tests/test_poetry.py]: Done in latest revision.

File: tests/test_poetry.py (hunk 3 above)
Comment (dlqqq) [reply tests/test_poetry.py]: Done in latest revision.

File: tests/test_poetry.py (hunk 4 above)
Comment (dlqqq) [reply tests/test_poetry.py]: Done in latest revision.

File: grayskull/strategy/pypi.py (hunk 5 above)
Comment (dlqqq) [reply grayskull/strategy/pypi.py:386]: Removed in latest revision.

File: grayskull/strategy/pypi.py (hunk 6 above)
Comment (dlqqq) [reply grayskull/strategy/pypi.py:583]: Done in latest revision.

File: tests/test_poetry.py (hunk 2 above)
Comment (dlqqq) [reply tests/test_poetry.py]: Done in latest revision.

File: grayskull/strategy/py_toml.py [hunk 7]
```diff
@@ -1,47 +1,212 @@
...
+            conda_version = encode_poetry_version(dep_spec)
+            run.append(f"{dep_name} {conda_version}")
+            continue
+
+        raise InvalidPoetryDependency(
+            "Expected Poetry dependency specification to be of type str or dict, "
+            f"received {type(dep_spec).__name__}"
+        )
+
+    return run, run_constrained
```
Comment (marcelotrevisani) [inline grayskull/strategy/py_toml.py:179]:
```suggestion
@singledispatch
def get_constrained_dep(dep_spec, dep_name):
    raise InvalidPoetryDependency(
        "Expected Poetry dependency specification to be of type str or dict, "
        f"received {type(dep_spec).__name__}"


@get_constrained_dep.register
def __get_constrained_dep_dict(dep_spec: Mapping, dep_name: str):
    conda_version = encode_poetry_version(dep_spec["version"])
    return f"{dep_name} {conda_version}"

@get_constrained_dep.register
def __get_constrained_dep_str(dep_spec: str, dep_name: str):
    conda_version = encode_poetry_version(dep_spec)
    return f"{dep_name} {conda_version}"

def encode_poetry_deps(poetry_deps: dict) -> Tuple[list, list]:
    run = []
    run_constrained = []
    for dep_name, dep_spec in poetry_deps.items():
        constrained_dep = get_constrained_dep(dep_spec, dep_name)
        try:
            assert dep_spec.get("optional", False)
            run_constrained.append(constrained_dep)
        except (AttributeError, AssertionError):
            run.append(constrained_dep)
    return run, run_constrained
```

**Comment by marcelotrevisani** [comment]
I have one more suggestion; after that, it should be good to go! :)

Thanks a lot for your contribution!

File: grayskull/strategy/py_toml.py (hunk 7 above)
Comment (dlqqq) [reply grayskull/strategy/py_toml.py]: Done in latest revision. 👍

**Review by marcelotrevisani — APPROVED** [review]

**Comment by marcelotrevisani** [comment]
Thanks for your contributions! 🚀

**Comment by marcelotrevisani** [comment]
sure, let me cut a new release

**Comment by xylar** [comment]
@dlqqq, I really appreciate this work.  The lack of `pyproject.toml` was becoming an increasing problem for me because I use grayskull as part of my workflow in maintaining a lot fo feedstocks.

**Pull request merged**